File: Project3B/lab3b.c
//NAME: Rohan Rao,Ethan Ngo
//EMAIL: raokrohan@gmail.com,ethan.ngo2019@gmail.com
//ID: 305339928,205416130

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

//for checking inode mode
#include <sys/stat.h>

//for time translation
#include <time.h>

//incudes for open
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

//includes for pread
#include <unistd.h>

//for errno
#include <errno.h>

//for ext2 structs
#include "ext2_fs.h"

//So we don't have to type struct every single time
typedef struct ext2_super_block ext2_super_block;
typedef struct ext2_group_desc ext2_group_desc;
typedef struct ext2_inode ext2_inode;

/*
    GLOBAL VARS GO HERE:
*/
unsigned int block_size;

unsigned int block_offset(unsigned int block)
{
    return 1024 + (block_size * (block - 1));
}

int is_block_used(int bno, char * bitmap)
{
    int index = 0, offset = 0;
    if (bno == 0)
        return 1;
    index = (bno - 1)/8; //which byte within the bitmap stores the info of this block#
    offset = (bno - 1)%8;
    return ((bitmap[index] & (1 << offset)) );
}

void epoch_to_gmt(char buf[], int bufsize, unsigned int t)
{
    time_t xtime = t;
    struct tm x_ts; 
    if(gmtime_r(&xtime, &x_ts) == NULL)
    {
        fprintf(stderr, "Error converting time...%s",strerror(errno));
        exit(2);
    }
    strftime(buf, bufsize, "%D %T", &x_ts);
}

int main(int argc, char* argv[])
{

    if(argc > 2 || argc == 1)
    {
        fprintf(stderr, "Incorrect arguments provided\n");
        fprintf(stderr, "Usage: %s FileSystemImageName.img \n", argv[0]);
        exit(1);
    }

    int fd = 0;
    if((fd = open(argv[1], O_RDONLY)) == -1)
    {
        fprintf(stderr, "Failed to open '%s'\n%s",argv[1], strerror(errno));
        exit(1);
    }

    //SUPERBLOCK DATA-------------
    unsigned int blocks_count = 0, inodes_count = 0, inode_size = 0, 
                blocks_per_group = 0, inodes_per_group = 0, first_non_res_inode = 0;
    block_size = 0;

    ext2_super_block super;

    if(pread(fd, &super, sizeof(super), 1024) == -1)
    {
        fprintf(stderr, "Problem with reading from fd...\n%s", strerror(errno));
        exit(1);
    }

    if(super.s_magic !=  EXT2_SUPER_MAGIC) 
    {
        fprintf(stderr, "Magic num != EXT2_SUPER_MAGIC...\n\
        File system may be corrupted or an invalid File System was given\n");
        exit(1);
    }

    blocks_count = super.s_blocks_count;
    inodes_count = super.s_inodes_count;
    block_size = 1024 << super.s_log_block_size; /* calculate block size in bytes */
    inode_size = super.s_inode_size;
    blocks_per_group = super.s_blocks_per_group;
    inodes_per_group = super.s_inodes_per_group;
    first_non_res_inode = super.s_first_ino;
    //END SUPERBLOCK DATA-------------


    //GROUP DESCRIPTORS DATA-------------
    unsigned int num_groups = blocks_count / blocks_per_group;
    if(!num_groups)
        num_groups = 1;

    ext2_group_desc* group_desc_table = 
    (ext2_group_desc*)malloc(num_groups*sizeof(ext2_group_desc));

    pread(fd, group_desc_table, num_groups*sizeof(ext2_group_desc), 1024 + block_size);
    //END GROUP DESCRIPTORS DATA-------------

    /*
    1.SUPERBLOCK *
    2.total number of blocks (decimal) *
    3.total number of i-nodes (decimal) *
    4.block size (in bytes, decimal) *
    5.i-node size (in bytes, decimal)
    6.blocks per group (decimal)
    7.i-nodes per group (decimal)
    8.first non-reserved i-node (decimal)
    */
    printf("SUPERBLOCK,%d,%d,%d,%d,%d,%d,%d\n", blocks_count,inodes_count,block_size,
                                                inode_size,blocks_per_group,
                                                inodes_per_group,first_non_res_inode);
    /*
    1.GROUP
    2.group number (decimal, starting from zero)
    3.total number of blocks in this group (decimal)
    4.total number of i-nodes in this group (decimal)
    5.number of free blocks (decimal)
    6.number of free i-nodes (decimal)
    7.block number of free block bitmap for this group (decimal)
    8.block number of free i-node bitmap for this group (decimal)
    9.block number of first block of i-nodes in this group (decimal)
    */
   unsigned int temp_blocks_per_group = 0;
   unsigned int temp_inodes_per_group = 0;
   for(unsigned int i = 0; i < num_groups; i++)
   {
        //account for lesser blocks in last group
        if(i == num_groups - 1)
           temp_blocks_per_group = blocks_count - blocks_per_group*(num_groups-1);
        //account for lesser inodes in last group
        if(i == num_groups - 1)
           temp_inodes_per_group = inodes_count - inodes_per_group*(num_groups-1);
           
        printf("GROUP,%d,%d,%d,%d,%d,%d,%d,%d\n", i, temp_blocks_per_group,temp_inodes_per_group,
                                            group_desc_table[i].bg_free_blocks_count,
                                            group_desc_table[i].bg_free_inodes_count,
                                            group_desc_table[i].bg_block_bitmap,
                                            group_desc_table[i].bg_inode_bitmap,
                                            group_desc_table[i].bg_inode_table);

        //BLOCK BITMAP------------------
        //1 = used
        //0 = free
        unsigned int block_bitmap;
        block_bitmap = group_desc_table[i].bg_block_bitmap;
        char* block_bits = (char*) malloc(block_size);
        
        pread(fd, block_bits, block_size, block_offset(block_bitmap));
        
        for(unsigned int j = 0; j < temp_blocks_per_group; j++)
            if(!is_block_used(j, block_bits))
                printf("BFREE,%d\n", j);
        //END BLOCK BITMAP---------------

        //INODE BITMAP------------------
        //1 = used
        //0 = free
        unsigned int inode_bitmap;
        inode_bitmap = group_desc_table[i].bg_inode_bitmap;
        char* inode_bits = (char*) malloc(block_size);
        pread(fd, inode_bits, block_size, block_offset(inode_bitmap));
        
        //made <= because first inode is reserved...
        //still need to iterate temp_inode_per_group times
        for(unsigned int j = 1; j <= temp_inodes_per_group; j++)
            if(!is_block_used(j, inode_bits))
                printf("IFREE,%d\n", j);
        //END INODE BITMAP---------------


        //INODE SUMMARY---------------
        /*
            1.INODE
            2.inode number (decimal)
            3.file type ('f' for file, 'd' for directory, 's' for symbolic link, '?" for anything else)
            4.mode (low order 12-bits, octal ... suggested format "%o")
            5.owner (decimal)
            6.group (decimal)
            7.link count (decimal)
            8.time of last I-node change (mm/dd/yy hh:mm:ss, GMT)
            9.modification time (mm/dd/yy hh:mm:ss, GMT)
            10/time of last access (mm/dd/yy hh:mm:ss, GMT)
            11.file size (decimal)
            12.number of (512 byte) blocks of disk space (decimal) taken up by this file
        */
        ext2_inode *inode_table = 
        (ext2_inode*)malloc(temp_inodes_per_group*sizeof(ext2_inode));
        pread(fd, inode_table, temp_inodes_per_group*sizeof(ext2_inode), block_offset(group_desc_table[i].bg_inode_table));

        for(unsigned int j = 1; j <= temp_inodes_per_group; j++)
            if(is_block_used(j, inode_bits))
            {
                ext2_inode cur_inode = inode_table[j];
                //only process if node and link count are non-zero
                if(cur_inode.i_mode == 0 || cur_inode.i_links_count == 0)
                    continue;
                //get filetype
                char filetype = '?';
                if(S_ISREG(cur_inode.i_mode))
                    filetype = 'f';
                if(S_ISDIR(cur_inode.i_mode))
                    filetype = 'd';
                if(S_ISLNK(cur_inode.i_mode))
                    filetype = 's';
                //get 12 lower bits in octal
                int octmask = (1<<12)-1;
                int octmode = cur_inode.i_mode & octmask;
                //format times (mm/dd/yy hh:mm:ss, GMT):
                char cbuf[30];
                char mbuf[30];
                char abuf[30];
                epoch_to_gmt(cbuf, 30, cur_inode.i_ctime);
                epoch_to_gmt(mbuf, 30, cur_inode.i_mtime);
                epoch_to_gmt(abuf, 30, cur_inode.i_atime);
                printf("INODE,%d,%c,%o,%d,%d,%d,%s,%s,%s,%d,%d",
                        j+1,filetype,octmode,cur_inode.i_uid,cur_inode.i_gid,cur_inode.i_links_count,cbuf,mbuf,abuf,
                        cur_inode.i_size,cur_inode.i_blocks);
                
                //print if a file, directory, or a symlink bigger than 60 bytes
                if(filetype == 'f' || filetype == 'd' || (filetype == 's' && cur_inode.i_size > 60))
                    for(int k = 0; k < 15; k++)
                        printf(",%u",cur_inode.i_block[k]);
                printf("\n");

                //DIRECTORY ENTRIES---------------
                /*
                1.DIRENT
                2.parent inode number (decimal) ... the I-node number of the directory that contains this entry
                3.logical byte offset (decimal) of this entry within the directory
                4.inode number of the referenced file (decimal)
                5.entry length (decimal)
                6.name length (decimal)
                7.name (string, surrounded by single-quotes). Don't worry about escaping, we promise there will 
                    be no single-quotes or commas in any of the file names.
                */
                
                struct ext2_dir_entry dir_entry;
                unsigned int iter;
                for(int k = 0; k < 12; k++)
                {
                    if(cur_inode.i_block[k] != 0 && filetype == 'd')
                    {
                        iter = 0;
                        while(iter < block_size) 
                        {
                            pread(fd, &dir_entry, sizeof(dir_entry), block_offset(cur_inode.i_block[k]) + iter);
                            if(dir_entry.inode != 0)
                            { 
                                memset(&dir_entry.name[dir_entry.name_len], 0, 256 - dir_entry.name_len);
                                printf("DIRENT,%d,%d,%d,%d,%d,'%s'\n",
                                j+1,iter,dir_entry.inode,dir_entry.rec_len, dir_entry.name_len, dir_entry.name);
                            }
                            iter += dir_entry.rec_len;
                        }
                    }
                }
                //END DIRECTORY ENTRIES---------------
                
                //INDIRECT BLOCK REFERENCES---------------
                /*
                1.INDIRECT
                2.I-node number of the owning file (decimal)
                3.(decimal) level of indirection for the block being scanned... 1 for single indirect, 2 for double indirect, 3 for triple
                4.logical block offset (decimal) represented by the referenced block. If the referenced block is a data block, this is the 
                logical block offset of that block within the file. If the referenced block is a single- or double-indirect block, this is 
                the same as the logical offset of the first data block to which it refers.
                5.block number of the (1, 2, 3) indirect block being scanned (decimal)... not the highest level block (in the recursive scan), 
                but the lower level block that contains the block reference reported by this entry.
                6.block number of the referenced block (decimal)
                */
                
                
                //single indirect block
                if(cur_inode.i_block[12] != 0)
                {
                    unsigned int* block_ptrs = malloc(block_size);
                    unsigned int num_ptrs = block_size/sizeof(unsigned int);

                    pread(fd, block_ptrs, block_size, block_offset(cur_inode.i_block[12]));
                    for(unsigned int k = 0; k < num_ptrs; k++)
                    {
                        struct ext2_dir_entry dir_entry;
                        unsigned int iter;
                        if(block_ptrs[k] != 0)
                        {
                            if(filetype == 'd')
                            {
                                iter = 0;
                                while(iter < block_size) 
                                {
                                    pread(fd, &dir_entry, sizeof(dir_entry), block_offset(block_ptrs[k]) + iter);
                                    if(dir_entry.inode != 0)
                                    { 
                                        memset(&dir_entry.name[dir_entry.name_len], 0, 256 - dir_entry.name_len);
                                        printf("DIRENT,%d,%d,%d,%d,%d,'%s'\n",
                                        j+1,iter,dir_entry.inode,dir_entry.rec_len, dir_entry.name_len, dir_entry.name);
                                    }
                                    iter += dir_entry.rec_len;
                                }
                            }

                            printf("INDIRECT,%d,%d,%d,%d,%d\n",j+1,1,k+12,cur_inode.i_block[12],block_ptrs[k]);
                        }
                    }
                    free(block_ptrs);
                }

                //double indirect block
                if(cur_inode.i_block[13] != 0)
                {
                    unsigned int* indir_block_ptrs = malloc(block_size);
                    unsigned int num_ptrs = block_size/sizeof(unsigned int);

                    pread(fd, indir_block_ptrs, block_size, block_offset(cur_inode.i_block[13]));
                    for(unsigned int k = 0; k < num_ptrs; k++)
                    {
                        //double indirect printout
                        if(indir_block_ptrs[k] != 0)
                            printf("INDIRECT,%d,%d,%d,%d,%d\n",j+1,2,256+k+12,cur_inode.i_block[13],indir_block_ptrs[k]);

                        //single indirect inside of double
                        unsigned int* block_ptrs = malloc(block_size);

                        pread(fd, block_ptrs, block_size, block_offset(indir_block_ptrs[k]));
                        for(unsigned int l = 0; l < num_ptrs; l++)
                        {
                            struct ext2_dir_entry dir_entry;
                            unsigned int iter;
                            if(block_ptrs[l] != 0)
                            {
                                if(filetype == 'd')
                                {
                                    iter = 0;
                                    while(iter < block_size) 
                                    {
                                        pread(fd, &dir_entry, sizeof(dir_entry), block_offset(block_ptrs[l]) + iter);
                                        if(dir_entry.inode != 0)
                                        { 
                                            memset(&dir_entry.name[dir_entry.name_len], 0, 256 - dir_entry.name_len);
                                            printf("DIRENT,%d,%d,%d,%d,%d,'%s'\n",
                                            j+1,iter,dir_entry.inode,dir_entry.rec_len, dir_entry.name_len, dir_entry.name);
                                        }
                                        iter += dir_entry.rec_len;
                                    }
                                }

                                printf("INDIRECT,%d,%d,%d,%d,%d\n",j+1,1,256+l+12,indir_block_ptrs[k],block_ptrs[l]);
                            }
                        }
                        free(block_ptrs);
                    }
                    free(indir_block_ptrs);
                }

                //triple indirect block
                if(cur_inode.i_block[14] != 0)
                {
                    unsigned int* indir_v2_block_ptrs = malloc(block_size);
                    unsigned int num_ptrs = block_size/sizeof(unsigned int);

                    pread(fd, indir_v2_block_ptrs, block_size, block_offset(cur_inode.i_block[14]));
                    for(unsigned int m = 0; m < num_ptrs; m++)
                    {
                        //double indirect printout
                        if(indir_v2_block_ptrs[m] != 0)
                            printf("INDIRECT,%d,%d,%d,%d,%d\n",j+1,3,65536+256+m+12,cur_inode.i_block[14],indir_v2_block_ptrs[m]);

                        unsigned int* indir_block_ptrs = malloc(block_size);
                        unsigned int num_ptrs = block_size/sizeof(unsigned int);

                        pread(fd, indir_block_ptrs, block_size, block_offset(indir_v2_block_ptrs[m]));
                        for(unsigned int k = 0; k < num_ptrs; k++)
                        {
                            //double indirect printout
                            if(indir_block_ptrs[k] != 0)
                                printf("INDIRECT,%d,%d,%d,%d,%d\n",j+1,2,65536+256+k+12,indir_v2_block_ptrs[m],indir_block_ptrs[k]);

                            //single indirect inside of double
                            unsigned int* block_ptrs = malloc(block_size);

                            pread(fd, block_ptrs, block_size, block_offset(indir_block_ptrs[k]));
                            for(unsigned int l = 0; l < num_ptrs; l++)
                            {
                                struct ext2_dir_entry dir_entry;
                                unsigned int iter;
                                if(block_ptrs[l] != 0)
                                {
                                    if(filetype == 'd')
                                    {
                                        iter = 0;
                                        while(iter < block_size) 
                                        {
                                            pread(fd, &dir_entry, sizeof(dir_entry), block_offset(block_ptrs[l]) + iter);
                                            if(dir_entry.inode != 0)
                                            { 
                                                memset(&dir_entry.name[dir_entry.name_len], 0, 256 - dir_entry.name_len);
                                                printf("DIRENT,%d,%d,%d,%d,%d,'%s'\n",
                                                j+1,iter,dir_entry.inode,dir_entry.rec_len, dir_entry.name_len, dir_entry.name);
                                            }
                                            iter += dir_entry.rec_len;
                                        }
                                    }

                                    printf("INDIRECT,%d,%d,%d,%d,%d\n",j+1,1,65536+256+l+12,indir_block_ptrs[k],block_ptrs[l]);
                                }
                            }
                            free(block_ptrs);
                        }
                        free(indir_block_ptrs);
                    }
                    free(indir_v2_block_ptrs);
                }
                //END INDIRECT BLOCK REFERENCES---------------
            }
        
        //END INODE SUMMARY---------------
   }

    exit(0);
}

File: Project3B/lab3b.py
#!/usr/bin/env python3

#NAME: Rohan Rao,Ethan Ngo
#EMAIL: raokrohan@gmail.com,ethan.ngo2019@gmail.com
#ID: 305339928,205416130

import sys
import csv

'''
PYTHON NOTES:

print("hello world", file=sys.stderr)
    To print to stderr do this


sys.argv[#] 
    to access command line args as you would in c

print(', '.join(row)) 
    to collapse a list into a string delimited by
    the specified string in ' '

global variablename
    use to reference and modify global variables inside functions
'''

#GLOBAL VARS--------------
exit_code = 0


#END GLOBAL VARS----------



class Superblock_Summary:
    def __init__(self):
        self.num_blocks = 0
        self.num_inodes = 0
        self.block_size = 0
        self.inode_size = 0
        self.blocks_per_group = 0
        self.inodes_per_group = 0
        self.f_nonres_inode = 0
    
    def initialize(self, nb, ni, bsz, isz, bpg, ipg, fni):
        self.num_blocks = nb
        self.num_inodes = ni
        self.block_size = bsz
        self.inode_size = isz
        self.blocks_per_group = bpg
        self.inodes_per_group = ipg
        self.f_nonres_inode = fni

class Group_Summary:
    def __init__(self):
        self.group_number = 0
        self.num_blocks = 0
        self.num_inodes = 0
        self.num_free_blocks = 0
        self.num_free_inodes = 0
        self.block_bitmap = 0
        self.inode_bitmap = 0
        self.f_inode_block = 0

    def initialize(self, gn, nb, ni, nfb, nfi, bb, ib, fib):
        self.group_number = gn
        self.num_blocks = nb
        self.num_inodes = ni
        self.num_free_blocks = nfb
        self.num_free_inodes = nfi
        self.block_bitmap = bb
        self.inode_bitmap = ib
        self.f_inode_block = fib

class Inode_Summary:
    def __init__(self):
        self.inode_number = 0
        self.file_type = 0
        self.mode = 0
        self.owner = 0
        self.group = 0
        self.link_count = 0
        self.time_create = 0
        self.time_modify = 0
        self.time_access = 0
        self.file_size = 0
        self.size_in_blocks = 0

    def initialize(self, inbr, ft, m, o, g, lc, tc, tm, ta, fs, sib):
        self.inode_number = inbr
        self.file_type = ft
        self.mode = m
        self.owner = o
        self.group = g
        self.link_count = lc
        self.time_create = tc
        self.time_modify = tm
        self.time_access = ta
        self.file_size = fs
        self.size_in_blocks = sib



def process_csv(filename):
    with open(filename, newline='') as csvfile:
        filereader = csv.reader(csvfile, delimiter=',')

        #putting data in list variable
        csvin = list()
        for row in filereader:
            csvin.append(row)

        #REFERENCES TO 'exit_code' NOW REFER TO GLOBAL 'exit_code'
        global exit_code

        it_temp = -1
        #OBJECTS FROM CSV DATA
        superblock = Superblock_Summary()
        group = Group_Summary()
        bfree_list = list()
        ifree_list = list()

        #HELPER OBJECTS 
        breserved_list = list()
        ireserved_list = list()
        
        unreferenced_blocks = list()
        referenced_blocks = list()
        dup_ref_blocks = dict()

        allocated_inodes = list()

        dirent_inode_references = dict()

        ptoc = dict() #parent to child directory
        ctop = dict() #child to parent directory

        #ADDING SUPERBLOCK AND GROUP DESCRIPTOR TABLE TO RESERVED AND REFERENCED LISTS
        breserved_list.extend([1, 2])
        referenced_blocks.extend([1,2])

        for row in csvin:
            it_temp += 1
            #print(row)
            #checks for basic csv validity, might not need
            if it_temp == 0 and row[0] != "SUPERBLOCK":
                print("Invalid csv: First entry not SUPERBLOCK...", file=sys.stderr)
                exit(2)
            if it_temp == 1 and row[0] != "GROUP":
                print("Invalid csv: Second entry not Group...", file=sys.stderr)
                exit(2)

            #POPULATING SUPERBLOCK OBJECT AND REFERENCE LISTS
            if row[0] == "SUPERBLOCK":
                superblock.initialize(int(row[1]), int(row[2]), int(row[3]), int(row[4]),
                                      int(row[5]), int(row[6]), int(row[7]))
                ireserved_list.extend(list(range(0, superblock.f_nonres_inode)))
                unreferenced_blocks.extend(list(range(0,superblock.num_blocks))) #TOOK OUT PLUS 1

            #POPULATING GROUP OBJECT AND REFERENCE LISTS
            if row[0] == "GROUP":
                group.initialize(int(row[1]), int(row[2]), int(row[3]), int(row[4]), 
                                 int(row[5]), int(row[6]), int(row[7]), int(row[8]))
                breserved_list.extend([group.block_bitmap, group.inode_bitmap, group.f_inode_block])
                referenced_blocks.extend([group.block_bitmap, group.inode_bitmap, group.f_inode_block])
                for _ in range(group.f_inode_block, group.f_inode_block + ((superblock.inode_size*superblock.inodes_per_group)//superblock.block_size)):
                    if _ not in referenced_blocks:
                        referenced_blocks.append(_)

            #POPULATING FREE BLOCK LIST
            if row[0] == "BFREE":
                if row[1] not in bfree_list:
                    bfree_list.append(int(row[1]))
            #POPULATING FREE INODE LIST
            if row[0] == "IFREE":
                if row[1] not in ifree_list:
                    ifree_list.append(int(row[1]))
            #POPULATING REFERENCED BLOCKS
            if row[0] == "INODE" and row[2] != "s":
                for _ in row[12:]:
                    if int(_) not in referenced_blocks:
                        referenced_blocks.append(int(_))
            if row[0] == "INDIRECT":
                if int(row[5]) not in referenced_blocks:
                        referenced_blocks.append(int(row[5]))
            #POPULATING RESERVED INODES
            if row[0] == "INODE":
                if int(row[1]) not in allocated_inodes:
                    allocated_inodes.append(int(row[1]))

            #RESERVED AND INVALID BLOCKS ALLOCATED
            if row[0] == "INODE" and row[2] != "s":
                temp_ = -1
                for _ in row[12:]:
                    temp_ += 1
                    ind_block = "BLOCK"
                    offset = temp_
                    if temp_ == 12:
                        ind_block = "INDIRECT BLOCK"
                        offset = 12
                    elif temp_ == 13:
                        ind_block = "DOUBLE INDIRECT BLOCK"
                        offset = 268
                    elif temp_ == 14:
                        ind_block = "TRIPLE INDIRECT BLOCK"
                        offset = 65804
                    if int(_) in breserved_list:
                        print("RESERVED", ind_block, _, "IN INODE", row[1], "AT OFFSET", offset)
                        exit_code = 2
                    if int(_) < 0 or int(_) > superblock.num_blocks:
                        print("INVALID", ind_block, _, "IN INODE", row[1], "AT OFFSET", offset)
                        exit_code = 2
                    
            #ALLOCATED BLOCK ON FREE LIST
            if row[0] == "INODE" and row[2] != "s":
                for _ in row[12:]:
                    if int(_) in bfree_list:
                        print("ALLOCATED BLOCK", _, "ON FREELIST")
                        exit_code = 2
            
            #SUM # OF BLOCK REFERENCES FOR EACH BLOCK
            if row[0] == "INODE" and row[2] != "s":
                    for _ in row[12:]:
                        if int(_) not in dup_ref_blocks and int(_) != 0:
                            dup_ref_blocks[int(_)] = 1
                        elif int(_) != 0:
                            dup_ref_blocks[int(_)] += 1

            #POPULATE INODE REFERENCES FROM DIRENTS
            if row[0] == "DIRENT":
                if int(row[3]) not in dirent_inode_references:
                    dirent_inode_references[int(row[3])] = 1
                else:
                    dirent_inode_references[int(row[3])] += 1

            #CUR '.' DIRECTORY VALIDITY CHECK
            if row[0] == "DIRENT":
                if row[6] == "'.'" and row[1] != row[3]:
                    print("DIRECTORY INODE", row[1], "NAME '.' LINK TO INODE", row[3], "SHOULD BE", row[1])
                    exit_code = 2
                elif row[6] == "'..'":
                    ptoc[int(row[1])] = int(row[3])
                elif row[6] != "'.'":
                    ctop[int(row[3])] = int(row[1])
                if int(row[1]) == 2 and row[6] == "'..'" and row[1] != row[3]:
                    print("DIRECTORY INODE", row[1], "NAME '..' LINK TO INODE", row[3], "SHOULD BE", row[1])
                    exit_code = 2
            #IN FOR LOOP----------
        #END FOR LOOP----------

        #DUPLICATE REFERENCES
        for row in csvin:
            if row[0] == "INODE":
                temp_ = -1
                for _ in row[12:]:
                    temp_ += 1
                    if int(_) in dup_ref_blocks and dup_ref_blocks[int(_)] > 1:
                        ind_block = "BLOCK"
                        block_num = int(_)
                        offset = temp_
                        if temp_ == 12:
                            ind_block = "INDIRECT BLOCK"
                            offset = 12
                        elif temp_ == 13:
                            ind_block = "DOUBLE INDIRECT BLOCK"
                            offset = 268
                        elif temp_ == 14:
                            ind_block = "TRIPLE INDIRECT BLOCK"
                            offset = 65804
                        print("DUPLICATE", ind_block, block_num, "IN INODE", row[1], "AT OFFSET", offset)
                        exit_code = 2

        # UNREFERENCED BLOCKS
        for _ in referenced_blocks:
            if _ in unreferenced_blocks:
                unreferenced_blocks.remove(_)
        for _ in bfree_list:
            if _ in unreferenced_blocks:
                unreferenced_blocks.remove(_)
        for _ in unreferenced_blocks:
            print("UNREFERENCED BLOCK", _)
            exit_code = 2

        #INODE ALLOCATION AUDITS
        for _ in ifree_list:
            if _ in allocated_inodes:
                print("ALLOCATED INODE", _, "ON FREELIST")
                exit_code = 2
        for _ in range(superblock.f_nonres_inode, group.num_inodes):
            if _ not in allocated_inodes and _ not in ifree_list:
                print("UNALLOCATED INODE", _, "NOT ON FREELIST")
                exit_code = 2

        #INODE LINK COUNT
        for row in csvin:
            if row[0] == "INODE":
                if int(row[1]) in dirent_inode_references:
                    if dirent_inode_references[int(row[1])] != int(row[6]):
                        print("INODE", row[1], "HAS", dirent_inode_references[int(row[1])], "LINKS BUT LINKCOUNT IS", row[6])
                        exit_code = 2
                elif int(row[6]) != 0:
                    print("INODE", row[1], "HAS 0", "LINKS BUT LINKCOUNT IS", row[6])
                    exit_code = 2
        
        #INVALID AND UNALLOCATED DIRENT REFERENCE TO INODE
        for row in csvin:
            if row[0] == "DIRENT":
                if int(row[3]) < 0 or int(row[3]) > group.num_inodes:
                    print("DIRECTORY INODE", row[1],"NAME", row[6], "INVALID INODE", row[3])
                    exit_code = 2
                elif int(row[3]) not in allocated_inodes:
                    print("DIRECTORY INODE", row[1],"NAME", row[6], "UNALLOCATED INODE", row[3])
                    exit_code = 2

        #VALIDATE PARENT AND CHILD REFERENCES
        #THEORETICALLY WORKS BUT THERE ARE NO TEST CASES FOR IT
        for p in ptoc:
            if p == 2:
                continue
            elif p not in ctop:
                print("DIRECTORY INODE", ptoc[p], "NAME '..' LINK TO INODE", p, " SHOULD BE", ptoc[p])
                exit_code = 2
            elif ptoc[p] != ctop[p]:
                print("DIRECTORY INODE", p + " NAME '..' LINK TO INODE " + p + " SHOULD BE " + ptoc[p])
                exit_code = 2
    #END 'process_csv'
def main():
    print("Running File System check on:", sys.argv[1], file=sys.stderr)
    process_csv(sys.argv[1])
    exit(exit_code)

if __name__ == "__main__":
    main()